src/detect-mqtt-flags.c

```diff
@@ -0,0 +1,365 @@
+    }
+
+    return de;
+
+error:
+    SCFree(de);
+    return NULL;
+}
+
+/**
+ * \internal
+ * \brief this function is used to add the parsed type query into the current signature
+ *
+ * \param de_ctx pointer to the Detection Engine Context
+ * \param s pointer to the Current Signature
+ * \param rawstr pointer to the user provided options
+ *
+ * \retval 0 on Success
+ * \retval -1 on Failure
+ */
+static int DetectMQTTFlagsSetup(DetectEngineCtx *de_ctx, Signature *s, const char *rawstr)
+{
+    DetectMQTTFlagsData *de = NULL;
+    SigMatch *sm = NULL;
+
+    de = DetectMQTTFlagsParse(rawstr);
+    if (de == NULL)
+        goto error;
+
+    sm = SigMatchAlloc();
+    if (sm == NULL)
+        goto error;
+
+    sm->type = DETECT_AL_MQTT_FLAGS;
+    sm->ctx = (SigMatchCtx *)de;
+
+    SigMatchAppendSMToList(s, sm, mqtt_flags_id);
+
+    return 0;
+
+error:
+    if (de) SCFree(de);
+    if (sm) SCFree(sm);
+    return -1;
+}
+
+/**
+ * \internal
+ * \brief this function will free memory associated with DetectMQTTFlagsData
+ *
+ * \param de pointer to DetectMQTTFlagsData
+ */
+void DetectMQTTFlagsFree(DetectEngineCtx *de_ctx, void *de_ptr)
+{
+    DetectMQTTFlagsData *de = (DetectMQTTFlagsData *)de_ptr;
+    if(de) SCFree(de);
+}
+
+/*
+ * ONLY TESTS BELOW THIS COMMENT
+ */
+
+#ifdef UNITTESTS
+/**
+ * \test MQTTFlagsTestParse01 is a test for a valid value
+ *
+ *  \retval 1 on success
+ *  \retval 0 on failure
+ */
+static int MQTTFlagsTestParse01 (void)
+{
+    DetectMQTTFlagsData *de = NULL;
+
+    de = DetectMQTTFlagsParse("retain");
+    FAIL_IF_NULL(de);
+    DetectMQTTFlagsFree(NULL, de);
+
+    de = DetectMQTTFlagsParse("dup");
+    FAIL_IF_NULL(de);
+    DetectMQTTFlagsFree(NULL, de);
+
+    de = DetectMQTTFlagsParse("retain,dup");
+    FAIL_IF_NULL(de);
+    DetectMQTTFlagsFree(NULL, de);
+
+    de = DetectMQTTFlagsParse("dup, retain");
+    FAIL_IF_NULL(de);
+    DetectMQTTFlagsFree(NULL, de);
+
+    PASS;
+}
+
+/**
+ * \test MQTTFlagsTestParse02 is a test for a valid value
+ *
+ *  \retval 1 on success
+ *  \retval 0 on failure
+ */
+static int MQTTFlagsTestParse02 (void)
+{
+    DetectMQTTFlagsData *de = NULL;
+    de = DetectMQTTFlagsParse("retain,!dup");
+    FAIL_IF_NULL(de);
+    DetectMQTTFlagsFree(NULL, de);
+
+    PASS;
+}
+
+/**
+ * \test MQTTFlagsTestParse03 is a test for an invalid value
+ *
+ *  \retval 1 on success
+ *  \retval 0 on failure
+ */
+static int MQTTFlagsTestParse03 (void)
+{
+    DetectMQTTFlagsData *de = NULL;
+    de = DetectMQTTFlagsParse("ref");
+    if (de) {
+        DetectMQTTFlagsFree(NULL, de);
+        FAIL;
+    }
+
+    PASS;
+}
+
+/**
+ * \test MQTTFlagsTestParse04 is a test for an invalid value
+ *
+ *  \retval 1 on success
+ *  \retval 0 on failure
+ */
+static int MQTTFlagsTestParse04 (void)
+{
+    DetectMQTTFlagsData *de = NULL;
+    de = DetectMQTTFlagsParse("dup,!");
+    if (de) {
+        DetectMQTTFlagsFree(NULL, de);
+        FAIL;
+    }
+
```

Comment:
What about a test with the signature "dup,!dup" ?

Reply:
Good idea. There's also code missing to detect this contradictory situation. Will add code to handle this case.